cmd/influxd/inspect/report_tsi/report_tsi.go

```diff
@@ -0,0 +1,481 @@
+
+	config := logger.NewConfig()
+	newLogger, err := config.New(os.Stderr)
+	if err != nil {
+		return err
+	}
+	report.sfile.Logger = newLogger
+
+	if err := report.sfile.Open(); err != nil {
+		return err
+	}
+	defer report.sfile.Close()
+
+	// Open all the indexes.
+	for id, pth := range report.shardPaths {
+		pth = path.Join(pth, "index")
+		// Verify directory is an index before opening it.
+		if ok, err := tsi1.IsIndexDir(pth); err != nil {
+			return err
+		} else if !ok {
+			return fmt.Errorf("not a TSI index directory: %q", pth)
+		}
+
+		report.shardIdxs[id] = tsi1.NewIndex(report.sfile,
+			"",
+			tsi1.WithPath(pth),
+			tsi1.DisableCompactions(),
+		)
+		if err := report.shardIdxs[id].Open(); err != nil {
+			return err
+		}
+		defer report.shardIdxs[id].Close()
+
+		// Initialise cardinality set to store cardinalities for this shard.
+		report.cardinalities[id] = map[string]*cardinality{}
+	}
+
+	// Calculate cardinalities of shards.
+	fn := report.cardinalityByMeasurement
+
+	// Blocks until all work done.
+	report.calculateCardinalities(fn)
+
+	// Print summary.
+	if err := report.printSummaryByMeasurement(); err != nil {
+		return err
+	}
+
+	allIDs := make([]uint64, 0, len(report.shardIdxs))
+	for id := range report.shardIdxs {
+		allIDs = append(allIDs, id)
+	}
+	sort.Slice(allIDs, func(i int, j int) bool { return allIDs[i] < allIDs[j] })
+
+	for _, id := range allIDs {
+		if err := report.printShardByMeasurement(id); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// calculateCardinalities calculates the cardinalities of the set of shard being
+// worked on concurrently. The provided function determines how cardinality is
+// calculated and broken down.
+func (report *reportTSI) calculateCardinalities(fn func(id uint64) error) error {
+	// Get list of shards to work on.
+	shardIDs := make([]uint64, 0, len(report.shardIdxs))
+	for id := range report.shardIdxs {
+		shardIDs = append(shardIDs, id)
+	}
+
+	errC := make(chan error, len(shardIDs))
+	var maxi uint32 // index of maximum shard being worked on.
+	for k := 0; k < report.concurrency; k++ {
+		go func() {
+			for {
+				i := int(atomic.AddUint32(&maxi, 1) - 1) // Get next partition to work on.
+				if i >= len(shardIDs) {
+					return // No more work.
+				}
+				errC <- fn(shardIDs[i])
+			}
+		}()
+	}
+
+	// Check for error
+	for i := 0; i < cap(errC); i++ {
+		if err := <-errC; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Cardinality struct and methods
+type cardinality struct {
+	name  []byte
+	short []uint32
+	set   *tsdb.SeriesIDSet
+}
+
+func (c *cardinality) add(x uint64) {
+	if c.set != nil {
+		c.set.AddNoLock(x)
+		return
+	}
+
+	c.short = append(c.short, uint32(x)) // Series IDs never get beyond 2^32
+
+	// Cheaper to store in bitmap.
+	if len(c.short) > useBitmapN {
+		c.set = tsdb.NewSeriesIDSet()
+		for i := 0; i < len(c.short); i++ {
+			c.set.AddNoLock(uint64(c.short[i]))
+		}
```

Comment:
More idiomatic would be to loop over the `range`:
```suggestion
		for _, s := range c.short {
			c.set.AddNoLock(uint64(s))
		}
```